src/kernel/mem/kheap.c

```diff
@@ -6,86 +6,194 @@
-                }
-                
-            } else {
-                KHeapMemoryNode *new_node = createKHeapNode(current_node, size);
-                if(prev_node != NULL){
-                    prev_node->next = new_node;
+
+    while( current_node != NULL ) {
+        // The size of a node contains also the size of the header, so when creating nodes we add headers
+        // We need to take it into account
+        size_t real_size = size + sizeof(KHeapMemoryNode);
+        real_size = align(real_size);
+        if( current_node->is_free) {
+            if( current_node->size >= real_size ) {
+                // Ok we have found a node big enough
+                if( current_node->size - real_size > KHEAP_MINIMUM_ALLOCABLE_SIZE ) {
+                    // We can keep shrinking the heap, since we still have enough space!
+                    // But we need a new node for the allocated area
+                    KHeapMemoryNode *new_node = create_kheap_node(current_node, real_size);
+                    current_node->is_free = false;
+                    current_node->size = real_size;
                 } else {
-                    kernel_heap_tail = new_node;
+                    // The current node space is not enough for shrinking, so we just need to mark the current_node as busy.
+                    // Size should not be touched.
+                    current_node->is_free = false;
+                    //current_node->size -= real_size;
                 }
+                return (void *) current_node + sizeof(KHeapMemoryNode);
             }
-            return (void *) current_node + sizeof(KHeapMemoryNode);
         }
+
+        if(current_node == kernel_heap_end) {
+            expand_heap(real_size);
+        }
+
         current_node = current_node->next;
     }
+}
 
-    if (size > KERNEL_PAGE_SIZE) { 
-        uint32_t number_of_pages = (size / KERNEL_PAGE_SIZE) + 1;
-        printf("KMALLOC SIZE > PAGE_SIZE, partially unsupported for now\n");
-        return  NULL;
-    } else {
-        uint64_t *new_address = ((void *)current_node) + current_node->size + sizeof(KHeapMemoryNode);
-        map_vaddress(new_address, 0);
+void expand_heap(size_t required_size) {
+    size_t number_of_pages = required_size / KERNEL_PAGE_SIZE + 1;
+    uint64_t heap_end = compute_kheap_end();
+    if( heap_end > end_of_mapped_memory ) {
+        //end_of_mapped memory marks the end of the memory mapped by the kernel loader.
+        //if the new heap address is above that, we need to map a new one, otherwise we can just mark it as used.
+        //That part temporary, it needs to be reviewed when the memory mapping will be reviewed.
+        map_vaddress_range(heap_end, 0, number_of_pages);
     }
+    KHeapMemoryNode *new_tail = (KHeapMemoryNode *) heap_end;
+    new_tail->next = NULL;
+    new_tail->prev = kernel_heap_end;
+    new_tail->size = KERNEL_PAGE_SIZE * number_of_pages;
+    new_tail->is_free = true;
+    
```

Comment:
Is it worth trying to merge `new_tail` and the current last node here (`kernel_heap_end` I think you call it)? It could help reduce fragmentation a little.

Reply:
and i think i should also upgrade kernel_heap_end to point to the new heap, and move the end... :)